### PR TITLE
ops-927 Added new data, corresponding to release on 10/10/2022

### PR DESCRIPTION
**Pull request by ami-day**
**Test data**

**Description**

Testing experiment design in which a cell suspension is used as input to another cell suspension (both in the "Cell suspension" tab).

E.g.
Mouse cells -> cell suspension 1
Human cells -> cell suspension 2
Mouse cells (cell suspension 1) + Human cells (cell suspension 2) -> cell suspension 3

### CODE REVIEW

**Review by amnonkhen — APPROVED** [review]

**Review by idazucchi — COMMENTED** [review]
The content of the spreadsheet is ok, but shouldn't the name of the spreadsheet stay the same?

Also this branch is behind master, is this a problem? 
I think this is the reason why some of the changes highlighted by the diff are part of the test data from the treatment protocol

**Comment by ami-day** [comment]
> The content of the spreadsheet is ok, but shouldn't the name of the spreadsheet stay the same?
> 
> Also this branch is behind master, is this a problem? I think this is the reason why some of the changes highlighted by the diff are part of the test data from the treatment protocol

Thanks for the review @idazucchi . I have re-uploaded the Schema_Test_Data.xlsx file with the appropriate same name. I also merged master into this branch, so it should now be updated to align with master.

**Review by idazucchi — APPROVED** [review]
Looks good!

**Review by ESapenaVentura — APPROVED** [review]
Please remember to upload the new files to the hca-util area with uuid `428794cc-cec2-4a98-8649-76accda2304f/`

**Comment by ami-day** [comment]
> Please remember to upload the new files to the hca-util area with uuid `428794cc-cec2-4a98-8649-76accda2304f/`

Yes will do this now :)

**Comment by ami-day** [comment]
Have requested review from dcp2 (here and slack)

**Review by ncalvanese1 — APPROVED** [review]

**Review by NoopDog — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
Before I can approve this, my team need to write a unit test against it in Azul. We'll discuss internally and get back to you with the timeline for that effort.

**Comment by ami-day** [comment]
> Before I can approve this, my team need to write a unit test against it in Azul. We'll discuss internally and get back to you with the timeline for that effort.

Hi @hannes-ucsc , would it be possible to get an update on this?

**Comment by hannes-ucsc** [comment]
@theathorn can answer the scheduling questions.

In the meantime, I would like to understand more about the graphs added in this PR. @ami-day, you mention three cell suspensions (CS). Could you post the UUIDs of those three CS?

> Mouse cells (cell suspension 1) + Human cells (cell suspension 2) -> cell suspension 3

If I understand that correctly, CS 3 is a mix of CS 1 and 2. Is that correct?

If so, can you point me at the graph edge or edges (post the UUIDs of the corresponding process entities) where that mixing occurs? I expect either one process with CS 1 and 2 as inputs and CS3 as the output, or two processes, one with CS1 as the input and one with CS2 the input, both with CS3 as output.

**Comment by ami-day** [comment]
> @theathorn can answer the scheduling questions.
> 
> In the meantime, I would like to understand more about the graphs added in this PR. @ami-day, you mention three cell suspensions (CS). Could you post the UUIDs of those three CS?
> 
> > Mouse cells (cell suspension 1) + Human cells (cell suspension 2) -> cell suspension 3
> 
> If I understand that correctly, CS 3 is a mix of CS 1 and 2. Is that correct?
> 
> If so, can you point me at the graph edge or edges (post the UUIDs of the corresponding process entities) where that mixing occurs? I expect either one process with CS 1 and 2 as inputs and CS3 as the output, or two processes, one with CS1 as the input and one with CS2 the input, both with CS3 as output.

Hi @hannes-ucsc, the cell suspension ids I gave you were a hypothetical example. The actual cell suspension ids and process ids and graph are as following:

**Cell suspension uuids**

AP1CS: e91df207-3754-423a-b7c3-4a52b6467b15
AP2CS: 45cce5ca-cd3d-4d54-bb19-e1e87add9ea0
AP3CS: a824c723-6307-4b4a-a0c7-d7bbb32f7756

**Process uuids**

1. AP1_specimen -> AP1CS: f7beae3a-d981-4014-b28c-f676a0836572
2. AP2_specimen -> AP2CS: 479e9452-ff89-4472-b7c3-17f09569ce6d
3. AP2CS -> AP3CS: fbf4dbe0-aec1-4d8a-96f4-9aaf7bb633a0

Process 3 listed above is the test case where a cell suspension is used as input to a cell suspension (AP2CS -> AP3CS)

Link to test data submission in staging from which i obtained the uuids: https://staging.contribute.data.humancellatlas.org/submissions/detail?uuid=d449543b-8415-46c9-8cbb-fbc20a4bd6cd&project=039a535d-a7a8-4808-9d7b-2b602c8bd8d1

Does this help?

**Comment by hannes-ucsc** [comment]
Now I am confused. The description states

> Mouse cells (cell suspension 1) + Human cells (cell suspension 2) -> cell suspension 3

and I asked in return 

> If I understand that correctly, CS 3 is a mix of CS 1 and 2. Is that correct?

As far as I can tell there was no explicit response to that question. 

However, in your latest post you write

> Process uuids
> AP1_specimen -> AP1CS: f7beae3a-d981-4014-b28c-f676a0836572
> AP2_specimen -> AP2CS: 479e9452-ff89-4472-b7c3-17f09569ce6d
> AP2CS -> AP3CS: fbf4dbe0-aec1-4d8a-96f4-9aaf7bb633a0

From which follows that AP3CS is derived only from AP2CS. Could you please help me resolve the apparent contradiction? What would really help is a diagram of the involved subgraphs. But since we're not even sure that we can support this in the near future (@theathorn decides schedule questions) I am hesitant to explicitly ask for that.

**Comment by theathorn** [comment]
Ami and Hannes, let's continue to resolve Hannes` questions on this ticket.
There are 2 datasets waiting to be submitted to the DCP that would require this support, but timeline is TBD.

**Comment by ami-day** [comment]
> Now I am confused. The description states
> 
> > Mouse cells (cell suspension 1) + Human cells (cell suspension 2) -> cell suspension 3
> 
> and I asked in return
> 
> > If I understand that correctly, CS 3 is a mix of CS 1 and 2. Is that correct?
> 
> As far as I can tell there was no explicit response to that question.
> 
> However, in your latest post you write
> 
> > Process uuids
> > AP1_specimen -> AP1CS: f7beae3a-d981-4014-b28c-f676a0836572
> > AP2_specimen -> AP2CS: 479e9452-ff89-4472-b7c3-17f09569ce6d
> > AP2CS -> AP3CS: fbf4dbe0-aec1-4d8a-96f4-9aaf7bb633a0
> 
> From which follows that AP3CS is derived only from AP2CS. Could you please help me resolve the apparent contradiction? What would really help is a diagram of the involved subgraphs. But since we're not even sure that we can support this in the near future (@theathorn decides schedule questions) I am hesitant to explicitly ask for that.

Hi @hannes-ucsc Sorry for the confusion, the original description was a **hypothetical** scenario, this should be ignored and only the updated description with the cell suspension uuids and process uuids should be considered. Thank you.

**Comment by hannes-ucsc** [comment]
> Sorry for the confusion, the original description was a hypothetical scenario, this should be ignored and only the updated description with the cell suspension uuids and process uuids should be considered. Thank you.

OK! I will henceforth ignore the contradictory original description, as well as anything the author describes as being "hypothetical".

In order to better understand this graph, and decide how to index it, I would need a diagram. Either my team could reverse engineer the metadata and create a diagram or the author could supply a diagram.

**Comment by ami-day** [comment]
> > Sorry for the confusion, the original description was a hypothetical scenario, this should be ignored and only the updated description with the cell suspension uuids and process uuids should be considered. Thank you.
> 
> OK! I will henceforth ignore the contradictory original description, as well as anything the author describes as being "hypothetical".
> 
> In order to better understand this graph, and decide how to index it, I would need a diagram. Either my team could reverse engineer the metadata and create a diagram or the author could supply a diagram.

Hi @hannes-ucsc I am sharing some slides with the graph images captured from the test data project in hca ingest staging: https://docs.google.com/presentation/d/1k9kMKvwL4PnjgcePauuBZE4o1M8BdhhQCT341MY99KU/edit#slide=id.p . I have shared it with your email address, please let me know if you have any problems accessing it.

I am also sharing the test data spreadsheet fyi: https://docs.google.com/spreadsheets/d/1ZPg59eDya6wUde7GX3GtfccNVKQFEcqsxumXfRUcX2k/edit#gid=310770490

Link to test data in ingest staging:
https://staging.contribute.data.humancellatlas.org/submissions/detail?uuid=d449543b-8415-46c9-8cbb-fbc20a4bd6cd&project=039a535d-a7a8-4808-9d7b-2b602c8bd8d1

If this is still unclear, I am happy to schedule a video call to screenshare.

**Comment by hannes-ucsc** [comment]
Thank you for the graphs. Unfortunately, the fact that they are artificially broken up into fragments by process, fragments that the viewer would have to mentally stitch together makes it really difficult to follow along. The individual process fragments are evident from the links file.

**Comment by ESapenaVentura** [comment]
Hi Hannes!

As Ami is leaving, I will pick it up from here!

So, if I understand this correctly, what you want is an accurate representation in the form of a diagram of the new data

I have prepared a very small png with that:

subgraph_cs_to_cs drawio

Basically, what this test data is creating is 2 new subgraphs. in reality, it's the same subgraph but it just changes the final output (analysis vs sequence files).

This subgraph represents a situation where a Cell suspension (APCS2) will be used to create another cell suspension (APCS3). In this specific case, they are linked together by a process that also has an enrichment protocol.

Let me know if this is what you want or if you want a little bit more detail into the subgraphs!

**Comment by hannes-ucsc** [comment]
Thank you, @ESapenaVentura. I've added what I believe are the implied edges to the graph. Let us know if I got that wrong.

203310393-51e8df3f-9c4d-41d1-8c6d-6b719ce9c8d8

> Basically, what this test data is creating is 2 new subgraphs. in reality, it's the same subgraph but it just changes the final output (analysis vs sequence files).

Is it fair to more accurately write that as 

> Basically, what this test data is creating is 2 new subgraphs. in reality, it's *two subgraphs of the same shape* but it just changes the final output (analysis vs sequence files).

Any other interpretation would confuse me. 


@danielsotirhos is working on a test case in the Azul code base.  He will reverse-engineer the graph structure and compare it with what's described here.

**Comment by ESapenaVentura** [comment]
> Thank you, @ESapenaVentura. I've added what I believe are the implied edges to the graph. Let us know if I got that wrong.
> 
> 203310393-51e8df3f-9c4d-41d1-8c6d-6b719ce9c8d8
> 
> > Basically, what this test data is creating is 2 new subgraphs. in reality, it's the same subgraph but it just changes the final output (analysis vs sequence files).
> 
> Is it fair to more accurately write that as
> 
> > Basically, what this test data is creating is 2 new subgraphs. in reality, it's _two subgraphs of the same shape_ but it just changes the final output (analysis vs sequence files).

Hi Hannes! Yes, that is exactly what this test data is representing

Many thanks for working in this

**Comment by dsotirho-ucsc** [comment]
@hannes-ucsc 
```
$ python scripts/view-bundle-outline.py -b 1ed68210-eaba-531d-ba9e-db80164d65ef
Project 90bf705c
Donor_organism 5bc2bf2f
+   Process 37e91a9f
+   +   Specimen_from_organism 3bee0c9a
+   +   +   Process c352fa85
+   +   +   +   Cell_suspension af3549f4
+   +   +   +   +   Process 9c7b713c
+   +   +   +   +   +   Cell_suspension 6ff38054
+   +   +   +   +   +   +   Process 1ed68210
+   +   +   +   +   +   +   +   Sequence_file de17aa7d
+   +   +   +   +   +   +   +   Sequence_file 8da4bdb9
```
```
$ python scripts/view-bundle-outline.py -b 7a330531-ec7f-5aee-84d4-2ba24d66e93b
Project 90bf705c
Donor_organism 5bc2bf2f
+   Process 37e91a9f
+   +   Specimen_from_organism 3bee0c9a
+   +   +   Process c352fa85
+   +   +   +   Cell_suspension af3549f4
+   +   +   +   +   Process 9c7b713c
+   +   +   +   +   +   Cell_suspension 6ff38054
+   +   +   +   +   +   +   Process 7a330531
+   +   +   +   +   +   +   +   Analysis_file caf6d2a8
```

**Comment by hannes-ucsc** [comment]
@ESapenaVentura, @danielsotirhos did some helpful analysis of the subgraphs added/modified here. There are two cell suspensions at play here. One is derived from the other. So far so good. Both cell suspensions have an estimated cell count of 20k. I assume that in reality there are only 20k cells involved. Please correct me if I am wrong.

**Comment by ESapenaVentura** [comment]
@hannes-ucsc both cell counts should be taken into account.

The biology behind this reasoning is that they took a cell suspension and divided it to generate another one, and later on they estimated the number of cells on each suspension

So, if the estimated_cell_count fields estimate 20k per cell suspension, the total would be 40k

Many thanks, and have a nice holiday!

**Comment by hannes-ucsc** [comment]
> they took a cell suspension and divided it to generate another one

This sentence confuses me. When I divide one thing I get at least two other things but in that sentence it appears that one thing is divided into one other thing. That seems counter-intuitive to me.

A cell suspension is an amount of fluid with cells suspended in it. One can transfer some of that fluid into another vessel (well or droplet) but then I would expect a proportional amount of the cells suspended in the original cell suspension to be transferred to the second cell suspension along with the transferred fluid. More importantly, I would also expect the overall amount of cells to remain constant, since cells are discrete entities and no cells are created and the loss of cells should be below a certain margin of error.

Then again, looking at the metadata graph it appears that an entire cell suspension is transformed wholly into another cell suspension and that only the result of the transformation is sequenced. But again, the overall number of discrete cells remains the same. No division takes place. No additional cells are created. So the number of sequenced cells is 20 thousand. I assumed that's really what we want the cell count display in the Data Browser to represent: the estimated number of sequenced or imaged cells.

**Comment by ESapenaVentura** [comment]
> Then again, looking at the metadata graph it appears that an entire cell suspension is transformed wholly into another cell suspension and that only the result of the transformation is sequenced. But again, the overall number of discrete cells remains the same. No division takes place. No additional cells are created. So the number of sequenced cells is 20 thousand. I assumed that's really what we want the cell count display in the Data Browser to represent: the estimated number of sequenced or imaged cells.

**tl;dr**: The number of cells in the `estimated_cell_count` field is accounted post-partition, not pre-partition

Slightly longer answer:

The graph I've shown in the previous comment is not the complete experimental graph - There are other subgraphs present in the experiment. In one of these subgraphs, that cell suspension is used as the input for an analysis file, `AP2_file.h5ad`. The cell suspension was used as an input to create the other cell suspension, and maybe it was enriched for something specific (In this specific scenario that Ami prepared, I see that it uses the same enrichment protocol, but it could have some other attached) and that differentiates it from the original cell suspension. 

However, what was left of the original cell suspension is also used for sequencing; hence, if the cell suspension states that there are 20k cells in there, what that means is that after the partition, there are 20k cells (Or, otherwise put, before the partition, there were 40k cells total in the suspension), and that is the number of cells expected for sequencing.

I understand that this could be modelled as "cell suspension A going into cell suspension B and cell suspension C", but that would only further complicate the experimental graph and would not add any additional useful information.

I hope I shed some light on the issue, let me know if you would like to discuss this further!

**Comment by hannes-ucsc** [comment]
The terms "divide" and "partition" keep coming up but I don't see this represented in the two subgraphs in this PR.

This is what we currently have in this PR (omitting processes, the colored boxes are subgraphs):

Blank diagram - Current

If the subgraphs were to reflect a "division" or "partition", they should include *two* derived cell suspensions. This is what that could look like:

Blank diagram - Desired(3)

Note that the upstream cell suspension has a cell count of 40k. The two derived cells suspensions have 20k cells each. 

The key take-away is that if, during the experiment, some of the original cell suspension was "syphoned off" into a new vessel, the remainder of the original cell suspension must be represented in the graph as a new cell suspension! If we don't do this, we will not be able to correctly count cells without resorting to hacks that will break in the future.

**Comment by ESapenaVentura** [comment]
You're right in those diagrams, that is exactly what we were trying to go for.

We were trying to make at least one of the subgraphs easier by skipping one of the cell suspensions, but I understand this will make it harder to the data portal for the cell counts so I will remake the test data to reflect this subgraph

I will ping for re-revision once it's done!

**Comment by ESapenaVentura** [comment]
@hannes-ucsc The subgraphs should be fixed now, there is now a cell suspension (CS2_pre, 40k cell count) that goes into CS2 and CS3

Let me know if there's anything else I can do!

**Comment by hannes-ucsc** [comment]
Thank you, @ESapenaVentura. @danielsotirhos could you please run your `view-bundle-outline` script on the most recent version and post output here?

**Comment by dsotirho-ucsc** [comment]
@hannes-ucsc 

The two bundles previously outlined (1ed68210 & 7a330531) are listed first.

```
$ python scripts/view-bundle-outline.py -b 1ed68210-eaba-531d-ba9e-db80164d65ef
Project 90bf705c
Donor_organism 5bc2bf2f
+   Process 37e91a9f
+   +   Specimen_from_organism 3bee0c9a
+   +   +   Process bd6a4a38
+   +   +   +   Cell_suspension 119c4135
+   +   +   +   +   Process b0092a47
+   +   +   +   +   +   Cell_suspension 6ff38054
+   +   +   +   +   +   +   Process 1ed68210
+   +   +   +   +   +   +   +   Sequence_file de17aa7d
+   +   +   +   +   +   +   +   Sequence_file 8da4bdb9
```

```
$ python scripts/view-bundle-outline.py -b 7a330531-ec7f-5aee-84d4-2ba24d66e93b
Project 90bf705c
Donor_organism 5bc2bf2f
+   Process 37e91a9f
+   +   Specimen_from_organism 3bee0c9a
+   +   +   Process bd6a4a38
+   +   +   +   Cell_suspension 119c4135
+   +   +   +   +   Process b0092a47
+   +   +   +   +   +   Cell_suspension 6ff38054
+   +   +   +   +   +   +   Process 7a330531
+   +   +   +   +   +   +   +   Analysis_file caf6d2a8
```

```
$ python scripts/view-bundle-outline.py -b d7b8cbff-aee9-5a05-a4a1-d8f4e720aee7
Project 90bf705c
Donor_organism 5bc2bf2f
+   Process 37e91a9f
+   +   Specimen_from_organism 3bee0c9a
+   +   +   Process bd6a4a38
+   +   +   +   Cell_suspension 119c4135
+   +   +   +   +   Process b8ec88d7
+   +   +   +   +   +   Cell_suspension af3549f4
+   +   +   +   +   +   +   Process d7b8cbff
+   +   +   +   +   +   +   +   Analysis_file f0c8ac7c
```

edit by @hannes-ucsc: I deleted the rest since they aren't modified by this PR.

**Comment by hannes-ucsc** [comment]
@ESapenaVentura is it intentional that there are now two h5ad files, one for each 20k cell suspension? If I understood your earlier comments correctly there is one cell suspension from which fastqz were made and another one from which an h5ad was made.

**Comment by ESapenaVentura** [comment]
> @ESapenaVentura is it intentional that there are now two h5ad files, one for each 20k cell suspension? If I understood your earlier comments correctly there is one cell suspension from which fastqz were made and another one from which an h5ad was made.

I did not add any new h5ad file, so if it wasn't showing up before it may have been a bug due to incorrectly splitting up the bundles. The structure showing up now is the intended one.

For what I can see in the spreadsheet, the only cell suspension that gets sequence files is AP3CS (The newly added one), getting both sequence files AND analysis files.

Before this test data, we didn't have any of the cell suspensions generating sequence files because it was an already tested subgraph (Normal donor --> specimen --> cell suspension --> sequence files) so we probably determined we didn't need to test for it.

Does it cause any issue?

**Comment by hannes-ucsc** [comment]
> Does it cause any issue?

The sequence files are fine, my question is pertaining analysis files. How many `.h5ad` files do you expect to be directly or indirectly derived from the primary cell suspension whose `biomaterial_id` is `AP2CS_pre`?

**Comment by ESapenaVentura** [comment]
> > Does it cause any issue?
> 
> The sequence files are fine, my question is pertaining analysis files. How many `.h5ad` files do you expect to be directly or indirectly derived from the primary cell suspension whose `biomaterial_id` is `AP2CS_pre`?

There are no h5ad files directly derived from AP2CS_pre, as it is not sequenced

There is one h5ad file derived from AP2CS and another one derived from AP3CS (And another one derived from AP1CS, but that one is unaffected on this iteration of test data). 

Would it be easier if we set-up a meeting to discuss this?

**Comment by hannes-ucsc** [comment]
OK, so no `.h5ad` files are derived directly from `AP2CS_pre` and two are derived indirectly. Yes?

**Comment by ESapenaVentura** [comment]
> OK, so no `.h5ad` files are derived directly from `AP2CS_pre` and two are derived indirectly. Yes?

Exactly!

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
See https://github.com/DataBiosphere/azul/issues/4425, PR https://github.com/DataBiosphere/azul/pull/4836 which should land in production next week or the week after.

**Comment by ESapenaVentura** [comment]
Thanks @hannes-ucsc ! We will hold the datasets that were pending on this for the next release then
